cpp/src/utilities/stream_pool.cpp

```diff
@@ -0,0 +1,205 @@
+   * streams, but this cannot be guaranteed if `count` is greater than the value returned by
+   * `get_stream_pool_size()`.
+   *
+   * This function is thread safe with respect to other calls to the same function.
+   *
+   * @param count The number of stream views to return.
+   * @return Vector containing `count` stream views.
+   */
+  virtual std::vector<rmm::cuda_stream_view> get_streams(uint32_t count) = 0;
+
+  /**
+   * @brief Get the number of stream objects in the pool.
+   *
+   * This function is thread safe with respect to other calls to the same function.
+   *
+   * @return the number of stream objects in the pool
+   */
+  virtual std::size_t get_stream_pool_size() const = 0;
+};
+
+/**
+ * @brief Implementation of `cuda_stream_pool` that wraps an `rmm::cuda_stram_pool`.
+ */
+class rmm_cuda_stream_pool : public cuda_stream_pool {
+  rmm::cuda_stream_pool _pool;
+
+ public:
+  rmm_cuda_stream_pool() : _pool{STREAM_POOL_SIZE} {}
+  rmm::cuda_stream_view get_stream() override { return _pool.get_stream(); }
+  rmm::cuda_stream_view get_stream(std::size_t stream_id) override
+  {
+    return _pool.get_stream(stream_id);
+  }
+
+  std::vector<rmm::cuda_stream_view> get_streams(uint32_t count) override
+  {
+    if (count > STREAM_POOL_SIZE) {
+      CUDF_LOG_WARN("get_streams called with count ({}) > pool size ({})", count, STREAM_POOL_SIZE);
+    }
+    auto streams = std::vector<rmm::cuda_stream_view>();
+    for (uint32_t i = 0; i < count; i++) {
+      streams.emplace_back(_pool.get_stream());
+    }
+    return streams;
+  }
+
+  std::size_t get_stream_pool_size() const override { return STREAM_POOL_SIZE; }
+};
+
+/**
+ * @brief Implementation of `cuda_stream_pool` that always returns `cudf::get_default_stream()`
+ */
+class debug_cuda_stream_pool : public cuda_stream_pool {
+ public:
+  rmm::cuda_stream_view get_stream() override { return cudf::get_default_stream(); }
+  rmm::cuda_stream_view get_stream(std::size_t stream_id) override
+  {
+    return cudf::get_default_stream();
+  }
+
+  std::vector<rmm::cuda_stream_view> get_streams(uint32_t count) override
+  {
+    return std::vector<rmm::cuda_stream_view>(count, cudf::get_default_stream());
+  }
+
+  std::size_t get_stream_pool_size() const override { return 1UL; }
+};
+
+/**
+ * @brief Initialize global stream pool.
+ */
+cuda_stream_pool* create_global_cuda_stream_pool()
+{
+  if (getenv("LIBCUDF_USE_DEBUG_STREAM_POOL")) return new debug_cuda_stream_pool();
+
+  return new rmm_cuda_stream_pool();
+}
+
+/**
+ * @brief RAII struct to wrap a cuda event and ensure it's proper destruction.
+ */
+struct cuda_event {
+  cuda_event()
+    : e_{[]() {
+        cudaEvent_t event;
+        CUDF_CUDA_TRY(cudaEventCreateWithFlags(&event, cudaEventDisableTiming));
+        return event;
+      }()}
+  {
+  }
+
+  operator cudaEvent_t() { return e_.get(); }
+
+ private:
+  struct deleter {
+    using pointer = cudaEvent_t;
+    auto operator()(cudaEvent_t e) { cudaEventDestroy(e); }
```

Comment:
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**💡 suggestion:** ‏ Should assert success here, so that you catch errors in a debug build. Can't throw because this is likely to be called in a destructor.